core/capabilities/remote/trigger_subscriber.go

```diff
@@ -106,78 +106,58 @@ func (s *triggerSubscriber) Info(ctx context.Context) (commoncap.CapabilityInfo,
-	if !ok {
-		regState = &subRegState{
-			callback:   make(chan commoncap.TriggerResponse, defaultSendChannelBufferSize),
-			rawRequest: rawRequest,
-		}
-		s.registeredWorkflows[request.Metadata.WorkflowID] = regState
-	} else {
-		regState.rawRequest = rawRequest
-		s.lggr.Warnw("RegisterTrigger re-registering trigger", "capabilityId", s.capInfo.ID, "donId", s.capDonInfo.ID, "workflowID", request.Metadata.WorkflowID)
+	responseChannel := s.responseChannels[request.Metadata.WorkflowID]
+	if responseChannel != nil {
+		close(responseChannel)
+		delete(s.responseChannels, request.Metadata.WorkflowID)
 	}
 
-	return regState.callback, nil
+	s.registrationClient.UnregisterWorkflow(request.Metadata.WorkflowID)
+	return nil
 }
 
-func (s *triggerSubscriber) registrationLoop() {
-	defer s.wg.Done()
-	ticker := time.NewTicker(s.config.RegistrationRefresh)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-s.stopCh:
-			return
-		case <-ticker.C:
-			s.mu.RLock()
-			s.lggr.Infow("register trigger for remote capability", "capabilityId", s.capInfo.ID, "donId", s.capDonInfo.ID, "nMembers", len(s.capDonInfo.Members), "nWorkflows", len(s.registeredWorkflows))
-			if len(s.registeredWorkflows) == 0 {
-				s.lggr.Infow("no workflows to register")
-			}
-			for _, registration := range s.registeredWorkflows {
-				// NOTE: send to all by default, introduce different strategies later (KS-76)
-				for _, peerID := range s.capDonInfo.Members {
-					m := &types.MessageBody{
-						CapabilityId:    s.capInfo.ID,
-						CapabilityDonId: s.capDonInfo.ID,
-						CallerDonId:     s.localDonInfo.ID,
-						Method:          types.MethodRegisterTrigger,
-						Payload:         registration.rawRequest,
-					}
-					err := s.dispatcher.Send(peerID, m)
-					if err != nil {
-						s.lggr.Errorw("failed to send message", "capabilityId", s.capInfo.ID, "donId", s.capDonInfo.ID, "peerId", peerID, "err", err)
-					}
-				}
-			}
-			s.mu.RUnlock()
-		}
-	}
+func (s *triggerSubscriber) isWorkflowRegistered(workflowID string) bool {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	_, found := s.responseChannels[workflowID]
+	return found
 }
 
-func (s *triggerSubscriber) UnregisterTrigger(ctx context.Context, request commoncap.TriggerRegistrationRequest) error {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+var errResponseChannelNotFound = errors.New("response channel not found")
 
-	state := s.registeredWorkflows[request.Metadata.WorkflowID]
-	if state != nil && state.callback != nil {
-		close(state.callback)
+func (s *triggerSubscriber) sendResponse(workflowID string, response commoncap.TriggerResponse) error {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	responseChannel, found := s.responseChannels[workflowID]
+	if !found {
+		return errResponseChannelNotFound
 	}
-	delete(s.registeredWorkflows, request.Metadata.WorkflowID)
-	// Registrations will quickly expire on all remote nodes.
-	// Alternatively, we could send UnregisterTrigger messages right away.
+
+	responseChannel <- response
```

Comment:
We do this under a lock - what if the reader is slow? It will block the whole subscriber object.